組件作業上傳

Frenkieli opened the pull request:
問題一、觸發順序問題

這邊去測試了 nextTick 、 computed 的set 和 beforeUpdate 的觸發順序

預想中 computed的set應該會先觸發更改資料，而後觸發刷新DOM前的beforeUpdate，最後在觸發DOM刷新結束的 nextTick

但實際測試確是 nextTick =>  computed : set => beforeUpdate ，nextTick沒有像文件所講的在DOM更新結束updated後才觸發

測試結果沒有在DOM觸發結束後才拿到 e.target.value 請問是 nextTick 的使用方式在哪邊出錯了嗎?

測試過程殘骸在檔案 vue-bike\src\components\searchBar.vue 下面

![觸發刷新結果](https://user-images.githubusercontent.com/49364993/90714867-3bf74f00-e2db-11ea-95b5-dd80f2c25499.jpg)

Frenkieli committed:
沒用過mitt先用mitt完成作業，感覺小型專案mitt比vuex好用好多喔！晚一點在拔掉mitt重寫

Frenkieli committed:
拔掉mitt用vue互傳重寫

Frenkieli committed:
refactor: 將子組件零件拉共用，將搜尋功能利用get.set重寫

將未來可能共用的組件拉出作全域共用(順便練習vue3怎麼註冊全域組件)，將搜尋功能以用get.set重寫感覺更加的直覺且迅速

  - vue-bike\src\main.js 練習VUE3如何註冊全域組件
  - vue-bike\src\components\UbikeTable.vue 使用全域組件
  - vue-bike\src\components\searchBar.vue 利用get.set重寫觸發機制，
這邊原本是利用用戶輸入enter或是點選搜尋的時候去觸發來規避畫面無法立即刷新的問題因為即便是使用$nextTick去等畫面刷新才觸發更新也沒有無法實現立即更新的功能(有bug)，但是使用get和set去觸發更新資料的時候就能很完美的實現即時搜尋的功能。

所以進行猜想測試所謂的DOM更新循環是不是指畫面要有更新，但input輸入的時候畫面並沒有更新所以function就立即執行所以拿到的是舊資料 ，這段側是因為輸入第二個字的時候畫面有刷新被否定了

而後測試set的觸發時間和nextTick的觸發時間，結果是先觸發nextTick console.log('刷新')過後才觸發computed的設定

猜想set的方法可能是在DOM被資料刷新完後才觸發，在此之前可能是先存在vue裡面的某個地方去刷畫面這樣

但利用beforeUpdate之後卻發現刷新和設定之後觸發才會觸發beforeUpdate裡面的console.log()

實際順序是　nextTick =>  computed : set => beforeUpdate　，好奇妙

Frenkieli committed:
test: 測試觸發順序

感覺跟文件說的不太一樣，直接問好了

kurotanshi commented:
> 問題一、觸發順序問題
> 
> 這邊去測試了 nextTick 、 computed 的set 和 beforeUpdate 的觸發順序
> 
> 預想中 computed的set應該會先觸發更改資料，而後觸發刷新DOM前的beforeUpdate，最後在觸發DOM刷新結束的 nextTick
> 
> 但實際測試確是 nextTick => computed : set => beforeUpdate ，nextTick沒有像文件所講的在DOM更新結束updated後才觸發
> 
> 測試結果沒有在DOM觸發結束後才拿到 e.target.value 請問是 nextTick 的使用方式在哪邊出錯了嗎?
> 
> 測試過程殘骸在檔案 vue-bike\src\components\searchBar.vue 下面
> 
> ![觸發刷新結果](https://user-images.githubusercontent.com/49364993/90714867-3bf74f00-e2db-11ea-95b5-dd80f2c25499.jpg)

Hi, 這個問題很有趣。

按照你的設定，我猜測你預想 `nextTick` 應該會在最後才執行對吧？ 但結果卻在一開始就出現了。

這裏我講一下我的看法。

你在 `input` 裡面同時用 `v-model` 綁定 `conditionName`，並且加入了 `keydown` 事件來執行 `searchStart` method：

``` html
<input 
  ref="searchName"
  id="searchName"
  class="form-control" 
  type="text" 
  v-model="conditionName"
  @keydown="searchStart"
>
```

這是 computed:
``` js
conditionName: {
  get: function({conditionNameProp}){
    return conditionNameProp;
  },
  set: function(val){
    let vm = this;
    vm.$emit('update:conditionNameProp', val);
    vm.$emit('page-change-event', 0);
    console.log(Date.now(), '設定');
  }
}
```

這是 method:
``` js
searchStart(){
  let vm = this;
  console.log('searchStart 觸發');

  vm.$nextTick(function(){
    console.log(Date.now(), 'searchStart nextTick 刷新', vm.$refs.searchName.value);
  })
}
```

結果當 input 被修改時， `keydown` 事件的 `searchStart` 先被執行，
並且印出 "searchStart 觸發"、連同 `nextTick` 的刷新也印出。

然後才進到 `computed` 的設定，最後才到 `beforeUpdate` hook 的 callback function。

這是因為 `$nextTick` 被調用的時候，會先去檢查「 DOM 操作相關的 queue」 是否還有未完成的任務，
如果有，則去等待 DOM 更新完成後執行，如果沒有就馬上執行。

跟 `setTimeout` 0 秒的情況不同的是，`setTimeout` 會等待所有非主線程的任務完成後才執行。

而真正寫入 data / DOM 其實是在 computed 的 set 而不是 methods。

最好的檢驗方式，就是我們將 `computed` 與 `methods` 分別加入 `setTimeout` 0 秒 與 `nextTick` 的操作：

``` js
// computed
conditionName: {
  get: function({conditionNameProp}){
    return conditionNameProp;
  },
  set: function(val){
    let vm = this;
    vm.$emit('update:conditionNameProp', val);
    vm.$emit('page-change-event', 0);
    console.log(Date.now(), '設定');

    // setTimeout 0 秒
    window.setTimeout(() => { console.log(Date.now(), 'computed - setTimeout'); }, 0);

    // nextTick
    vm.$nextTick(function(){
      console.log(Date.now(), 'computed-set nextTick 刷新', vm.$refs.searchName.value);
    });
  }
```

``` js
// methods
searchStart () {
  let vm = this;
  console.log('searchStart 觸發');

  // setTimeout 0 秒
  window.setTimeout(() => { console.log(Date.now(), 'searchStart - setTimeout'); }, 0);

  // nextTick
  vm.$nextTick(function(){
    console.log(Date.now(), 'searchStart nextTick 刷新', vm.$refs.searchName.value);
  });
}
},
```

在 input 被更新之後，會依序出現這樣的結果：
<img width="671" alt="截圖 2020-08-20 下午3 50 44" src="https://user-images.githubusercontent.com/362912/90732202-56422480-e2fd-11ea-86a5-208c0178cd20.png">


在 method 裡的 nextTick 會先被執行，然後進到 computed 、 beforeUpdate ，
等待 DOM 同步完成後執行 computed 的 nextTick，最後才是兩者的 setTimeout。

希望有幫助到你對 nextTick 順序的觀念理解 :)

Frenkieli committed:
test: 測試 nextTick 觸發的時間

從老師解答後進行在測試

結論： nextTick 被觸發的瞬間會去檢查Vue的　"資料"　是否有在等待變動畫面DOM，等到資料變動畫面DOM的所有事件結束後在執行 nextTick 內的function
